azurelinuxagent/ga/exthandlers.py

```diff
@@ -276,6 +288,78 @@ def run(self):
+        except Exception as e:
+            msg = u"Exception retrieving VmArtifactsProfile blob: {0}".format(ustr(e))
+            detailed_msg = '{0} {1}'.format(msg, traceback.extract_tb(get_traceback(e)))
+            logger.warn(msg)
+
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.ExtensionProcessing,
+                      is_success=False,
+                      message=detailed_msg)
+
+    def process_goal_state(self, incarnation, sequence_number):
+        fabric_goal_state_changed = self.determine_fabric_changed(incarnation)
+        fast_track_changed = self.determine_fast_track_changed(sequence_number)
+        if fabric_goal_state_changed:
+            self.protocol.set_fast_track(False)
+
+        # Fabric goal state takes precedence over VMArtifactsProfile blob goal state
+        # So, process the Fabric goal state first
+        self.handle_ext_handlers(fabric_goal_state_changed)
+        self.last_incarnation = incarnation
+
+        # Now process the VMArtifactsProfile if it changed. Note that just because the VMArtifactsProfile changed
+        # doesn't mean the FastTrack goal state changed
+        if fast_track_changed and self.artifacts_profile is not None:
+            vm_artifacts_ext_conf = self.artifacts_profile.transform_to_extensions_config()
+            self.process_vm_artifacts_profile_for_fast_track(sequence_number, vm_artifacts_ext_conf)
+        elif self.protocol.client.get_use_fast_track():
+            # Our last goal state was via VMArtifactsProfile, so reset the extension handlers
+            self.ext_conf = self.protocol.client.get_ext_conf(read_always=True)
+            self.ext_handlers = self.ext_conf.ext_handlers
+
+    def determine_fabric_changed(self, incarnation):
+        fabric_goal_state_changed = False
+
+        if self.last_incarnation != incarnation:
+            logger.info('Goal state changed with incarnation={0}(last:{1})'.
+                        format(str(incarnation), str(self.last_incarnation)))
+            fabric_goal_state_changed = True
+
+        return fabric_goal_state_changed
+
+    def determine_fast_track_changed(self, sequence_number):
+        fast_track_changed = False
+
+        if sequence_number is not None and self.last_vm_artifacts_seqno != sequence_number and \
+                conf.get_extensions_fast_track_enabled():
+            logger.info('VMArtifactsProfile changed with sequence number={0}(last:{1})'.
+                        format(str(sequence_number), str(self.last_vm_artifacts_seqno)))
+            fast_track_changed = True
+        return fast_track_changed
+
+    def process_vm_artifacts_profile_for_fast_track(self, sequence_number, vm_artifacts_ext_conf):
+        if vm_artifacts_ext_conf is not None:
+            # We need to add the status upload blob from Fabric goal state because we don't receive this
```

Comment:
I have concerns about this, what would happen if status blob is updated but there was no new fabric GoalState generated so `self.ext_conf.status_upload_blob` would point to the old status blob and not the new one and then we get a new ExtensionConfig through fasttrack?

Reply:
Can we not make it more resilient by adding the status blob as part of the ArtifactBlob we get from FastTrack too to ensure that its always the latest?